internal/servers/controller/handlers/authmethods/oidc.go

```diff
@@ -123,6 +139,99 @@ func (s Service) updateOidcInRepo(ctx context.Context, scopeId string, req *pbs.
 	return out, nil
 }
 
+func (s Service) authenticateOidc(ctx context.Context, req *pbs.AuthenticateRequest, authResults *auth.VerifyResults) (*pbs.AuthenticateResponse, error) {
+	const op = "authmethod_service.(Service).authenticateOidc"
+	if req == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "nil request")
+	}
+	if authResults == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "nil auth results")
+	}
+	switch req.GetCommand() {
+	case startCommand:
+		return s.authenticateOidcStart(ctx, req, authResults)
+	}
+
+	// Default is tokenCommand -- note we've already checked that it's one of
+	// these three in the validation function
+	// TODO
+	return &pbs.AuthenticateResponse{Command: req.GetCommand(), Attributes: nil}, nil
+}
+
+func (s Service) authenticateOidcStart(ctx context.Context, req *pbs.AuthenticateRequest, authResults *auth.VerifyResults) (*pbs.AuthenticateResponse, error) {
+	const op = "authmethod_service.(Service).authenticateOidcStart"
+	if req == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "nil request")
+	}
+	if authResults == nil {
+		return nil, errors.New(errors.InvalidParameter, op, "nil auth results")
+	}
+
+	var opts []oidc.Option
+	if req.GetAttributes() != nil && req.GetAttributes().GetFields() != nil {
+		if val, ok := req.GetAttributes().GetFields()[cachedRoundtripPayloadField]; ok {
+			opts = append(opts, oidc.WithRoundtripPayload(val.GetStringValue()))
+		}
+	}
+
+	authUrl, tokenUrl, tokenId, err := oidc.StartAuth(ctx, s.oidcRepoFn, req.GetAuthMethodId(), opts...)
+	if err != nil {
+		// TODO: Log something
+		return nil, errors.New(errors.Internal, op, "Error generating parameters for starting the OIDC flow.")
+	}
+
+	attrsMap := map[string]interface{}{
+		authUrlField:  authUrl.String(),
+		tokenUrlField: tokenUrl.String(),
+		tokenIdField:  tokenId,
+	}
+
+	attrs, err := structpb.NewStruct(attrsMap)
+	if err != nil {
+		return nil, errors.New(errors.Internal, op, "Error marshaling parameters.")
+	}
+
+	return &pbs.AuthenticateResponse{Command: req.GetCommand(), Attributes: attrs}, nil
+}
+
+func validateAuthenticateOidcRequest(req *pbs.AuthenticateRequest) error {
+	badFields := make(map[string]string)
+
+	switch req.GetCommand() {
+	case startCommand:
+		if req.GetAttributes() != nil && req.GetAttributes().GetFields() != nil {
```

Comment:
I would imagine a good validation of an authenticate request is that the attributes are able to be converted to the proto representation (which validates that all the fields passed in are recognizable).  That would make it easier to validate the received values.

Reply:
There isn't a proto for this. We could make one, but I'm not sure that's the right thing to do, because this field is defined as an arbitrary struct. So you'd just convert the struct into a proto containing.....a struct, and you'd have the same logic one level down.

It's also not clear to me that ensuring all fields passed in are recognizable is a good thing. It's nice to not accept fields if we don't know about them. On the other hand, not doing so can create complex upgrading issues later on due to breaking clients that pass parameters that older servers don't know about.